### PR TITLE
MSI-783: Add service GetProductTypeBySku

### PR DESCRIPTION
**Pull request by serhii-balko**
### Fixed Issues (if relevant)
1. magento-engcom/msi#783: Add core service GetProductTypeBySku

### CODE REVIEW

**Review by naydav — CHANGES_REQUESTED** [review]
Alos, we need to apply created services to

https://github.com/magento-engcom/msi/pull/622/files#diff-cb68a89e0c4deb4f3e65596ed1232028

https://github.com/magento-engcom/msi/pull/622/files#diff-fe8392cfd690a7369b4e5cea9e1cd977

https://github.com/magento-engcom/msi/pull/622/files#diff-0483ec9308884bfd8cbeb256701c3c64

We can do it in this PR

**Comment by serhii-balko** [comment]
@naydav done

File: app/code/Magento/InventoryCatalog/Plugin/CatalogInventory/UpdateSourceItemAtLegacyStockItemSavePlugin.php [hunk 1]
```diff
@@ -90,18 +99,37 @@ public function aroundSave(ItemResourceModel $subject, callable $proceed, Abstra
...
         return $typeId;
     }
+
+    /**
+     * Get product sku by id
+     *
+     * @param int $productId
+     * @return string|null
+     */
+    private function getProductSkuById(int $productId)
```
Comment (naydav) [inline app/code/Magento/InventoryCatalog/Plugin/CatalogInventory/UpdateSourceItemAtLegacyStockItemSavePlugin.php]: Pls use
\Magento\InventoryCatalog\Model\GetProductIdsBySkusInterface

File: app/code/Magento/InventoryCatalog/Plugin/CatalogInventory/UpdateSourceItemAtLegacyStockItemSavePlugin.php [hunk 2]
```diff
@@ -90,18 +99,37 @@ public function aroundSave(ItemResourceModel $subject, callable $proceed, Abstra
...
         }
 
         return $typeId;
     }
+
+    /**
+     * Get product sku by id
+     *
+     * @param int $productId
+     * @return string|null
```
Comment (naydav) [inline app/code/Magento/InventoryCatalog/Plugin/CatalogInventory/UpdateSourceItemAtLegacyStockItemSavePlugin.php]: We don't have cases when the method returns `null`
See implementation of interface `\Magento\InventoryCatalog\Model\GetProductIdsBySkusInterface`

File: app/code/Magento/InventorySales/Plugin/CatalogInventory/StockManagement/ProcessRegisterProductsSalePlugin.php [hunk 3]
```diff
@@ -85,19 +103,26 @@ public function aroundRegisterProductsSale(StockManagement $subject, callable $p
...
+
         $stockId = (int)$this->stockByWebsiteIdResolver->get((int)$websiteId)->getStockId();
         $productSkus = $this->getSkusByProductIds->execute(array_keys($items));
-
         $this->checkItemsQuantity($items, $productSkus, $stockId);
+
+        $productTypes = $this->getProductTypesBySkus->execute(array_values($productSkus));
         $reservations = [];
         foreach ($productSkus as $productId => $sku) {
+            if (!$this->isSourceItemsAllowedForProductType->execute($productTypes[$sku])) {
```
Comment (naydav) [inline app/code/Magento/InventorySales/Plugin/CatalogInventory/StockManagement/ProcessRegisterProductsSalePlugin.php]: Why do we need this check?
Do you discover situation when passed sku is related to complex products?

Comment (serhii-balko) [reply app/code/Magento/InventorySales/Plugin/CatalogInventory/StockManagement/ProcessRegisterProductsSalePlugin.php]: The issue about this already exists
https://github.com/magento-engcom/msi/issues/790
@nmalevanec working on it

File: app/code/Magento/InventoryCatalog/Plugin/CatalogInventory/UpdateSourceItemAtLegacyStockItemSavePlugin.php (hunk 1 above)
Comment (serhii-balko) [reply app/code/Magento/InventoryCatalog/Plugin/CatalogInventory/UpdateSourceItemAtLegacyStockItemSavePlugin.php]: In this case i have product id and i need to know product sku.
The proposed service can not do this

Comment (naydav) [reply app/code/Magento/InventoryCatalog/Plugin/CatalogInventory/UpdateSourceItemAtLegacyStockItemSavePlugin.php]: `GetSkusByProductIdsInterface`

File: app/code/Magento/InventoryCatalog/Plugin/CatalogInventory/UpdateSourceItemAtLegacyStockItemSavePlugin.php (hunk 2 above)
Comment (serhii-balko) [reply app/code/Magento/InventoryCatalog/Plugin/CatalogInventory/UpdateSourceItemAtLegacyStockItemSavePlugin.php]: fixed

File: app/code/Magento/InventoryCatalog/Plugin/CatalogInventory/UpdateSourceItemAtLegacyStockItemSavePlugin.php (hunk 1 above)
Comment (serhii-balko) [reply app/code/Magento/InventoryCatalog/Plugin/CatalogInventory/UpdateSourceItemAtLegacyStockItemSavePlugin.php]: done

**Pull request merged**